### PR TITLE
Decouple API representations from internal models

### PR DESCRIPTION
**Pull request by ConnorDoyle**
This is a start towards fixing #1505 so we can start experimenting with a v3 API.

_TODO:_

- [x] New V2AppDefinition for REST API use
- [x] New V2Group for REST API use
- [x] Move AppUpdate to the `mesosphere.marathon.api.v2.json` package
- [x] Move GroupUpdate to the `mesosphere.marathon.api.v2.json` package
- [x] Make sure integration tests still pass

_FUTURE WORK:_

- [ ] Fix AppUpdate usage outside the `mesosphere.marathon.api.v2` package
- [ ] Fix GroupUpdate usage outside the `mesosphere.marathon.api.v2` package
- [ ] Track event bus subscribers by version, form requests appropriately (can be postponed)
- [ ] Clean up and normalize model validation logic (can be postponed)

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/api/v2/json/V2AppDefinition.scala [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    * port mappings.
+    */
+  def portIndicesAreValid(): Boolean =
+    this.toAppDefinition().portIndicesAreValid()
+
+  /**
+    * Returns the canonical internal representation of this API-specific
+    * application defintion.
+    */
+  def toAppDefinition(): AppDefinition =
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/json/V2AppDefinition.scala]: I would prefer this method without `()` since it has no side effects.

File: src/main/scala/mesosphere/marathon/api/v2/json/V2AppDefinition.scala [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    portIndicesAreValid(),
+    "Health check port indices must address an element of the ports array or container port mappings."
+  )
+
+  /**
+    * Returns true if all health check port index values are in the range
+    * of ths app's ports array, or if defined, the array of container
+    * port mappings.
+    */
+  def portIndicesAreValid(): Boolean =
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/json/V2AppDefinition.scala]: without `()`?

**Comment by ConnorDoyle** [comment]
@kolloch ready for review.

**Comment by kolloch** [comment]
Hmmm. Looks like it needs a rebase.

**Comment by ConnorDoyle** [comment]
@kolloch likely; this touches lots of files so any interleaved commits to master are liable to require rebase.  Doing that now.  Meanwhile, any other comments about the change set?

**Comment by ConnorDoyle** [comment]
Rebased.

**Comment by ConnorDoyle** [comment]
@kolloch are we waiting until the `0.9.0` RC cycle is complete to move forward with this?

File: src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala [hunk 3]
```diff
@@ -218,15 +224,17 @@ class AppsResource @Inject() (
...
     BeanValidation.requireValid(ModelValidation.checkUpdates(updates))
     val version = Timestamp.now()
-    def updateApp(update: AppUpdate, app: AppDefinition): AppDefinition = {
-      update.version.flatMap(v => service.getApp(app.id, v)).orElse(Some(update(app))).getOrElse(app)
+    def updateApp(update: V2AppUpdate, app: AppDefinition): AppDefinition = {
+      val previousVersion = update.version.flatMap(service.getApp(app.id, _))
+      previousVersion.orElse(Some(update(V2AppDefinition(app)).toAppDefinition)).getOrElse(app)
     }
     def updateGroup(root: Group): Group = updates.foldLeft(root) { (group, update) =>
+      val transform = (app: AppDefinition) => updateApp(update, app)
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala]: why not a def?

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala [hunk 4]
```diff
@@ -389,89 +389,126 @@ trait AppDefinitionFormats {
...
-        "container" -> app.container,
-        "healthChecks" -> app.healthChecks,
-        "dependencies" -> app.dependencies,
-        "upgradeStrategy" -> app.upgradeStrategy,
-        "labels" -> app.labels,
-        "acceptedResourceRoles" -> app.acceptedResourceRoles,
-        "version" -> app.version
-      )
+    Writes[V2AppDefinition] { app =>
+      try {
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: Why did you add the try again?

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala [hunk 5]
```diff
@@ -389,89 +389,126 @@ trait AppDefinitionFormats {
...
+      }
+      catch {
+        case NonFatal(e) =>
+          throw e
+      }
     }
   }
 
-  implicit lazy val AppUpdateReads: Reads[AppUpdate] = {
+  implicit val WithTaskCountsAndDeploymentsWrites: Writes[WithTaskCountsAndDeployments] =
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: the vals above are all lazy to prevent problems if you access unitialized vals in your Format indirectly. I like that because it is easy to make a mistake otherwise.

File: src/main/scala/mesosphere/marathon/api/v2/json/V2AppDefinition.scala [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    @JsonProperty
+    val tasksUnhealthy: Int = healthCounts.unhealthy
+
+    /**
+      * Snapshot of the running deployments that affect this app
+      */
+    @JsonProperty
+    def deployments: Seq[Identifiable] = {
+      runningDeployments.collect {
+        case plan: Any if plan.affectedApplicationIds contains app.id => Identifiable(plan.id)
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/json/V2AppDefinition.scala]: Why did you change this from `plan: DeploymentPlan` to `plan: Any`? To be honest, I am surprised that this compiles.

Not related to this PR but out of curiosity: Why are we passing the runningDeployments to all AppDefinitions and filter them afterwards?

File: src/main/scala/mesosphere/marathon/api/v2/json/V2DeploymentPlan.scala [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+final case class V2DeploymentPlan(
+  id: String,
+  original: V2Group,
+  target: V2Group,
+  steps: Seq[DeploymentStep],
+  version: Timestamp)
+
+object V2DeploymentPlan {
+  def apply(plan: DeploymentPlan): V2DeploymentPlan =
+    V2DeploymentPlan(
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/json/V2DeploymentPlan.scala]: named arguments would be great.

File: src/main/scala/mesosphere/marathon/api/v2/json/V2Group.scala [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    * group.
+    */
+  def toGroup(): Group =
+    Group(
+      id,
+      apps.map(_.toAppDefinition),
+      groups.map(_.toGroup),
+      dependencies,
+      version
+    )
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/json/V2Group.scala]: Named arguments?

File: src/main/scala/mesosphere/marathon/api/v2/json/V2Group.scala [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  val defaultApps: Set[V2AppDefinition] = Set.empty
+  val defaultGroups: Set[V2Group] = Set.empty
+
+  def apply(group: Group): V2Group =
+    V2Group(
+      group.id,
+      group.apps.map(V2AppDefinition(_)),
+      group.groups.map(V2Group(_)),
+      group.dependencies,
+      group.version)
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/json/V2Group.scala]: Named arguments?

File: src/main/scala/mesosphere/marathon/api/v2/json/V2GroupUpdate.scala [hunk 10]
```diff
@@ -33,17 +33,17 @@ case class GroupUpdate(
         .map(update => update.toGroup(update.groupId.canonicalPath(current.id), timestamp))
       groupUpdates.toSet ++ groupAdditions
     }
-    val effectiveApps = apps.getOrElse(current.apps).map(toApp(current.id, _, timestamp))
+    val effectiveApps: Set[V2AppDefinition] = apps.getOrElse(current.apps).map(toApp(current.id, _, timestamp))
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/json/V2GroupUpdate.scala]: I like that you added the type here :)

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala [hunk 11]
```diff
@@ -331,80 +306,4 @@ object AppDefinition {
...
-        app.backoffFactor, app.maxLaunchDelay, app.container,
-        app.healthChecks, app.dependencies, app.upgradeStrategy,
-        app.labels, app.acceptedResourceRoles, app.version) {
-
-    /**
-      * Snapshot of the number of staged (but not running) tasks
-      * for this app
-      */
-    @JsonProperty
-    val tasksStaged: Int = appTasks.count { eTask =>
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: I think that this property as a lazy val or something like that is generally useful. Not only in the API.

But of course, we can recreate it if we actually use it.

File: src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala [hunk 12]
```diff
@@ -315,34 +323,3 @@ class AppsResource @Inject() (
...
-
-object AppsResource {
-  implicit val WithTaskCountsAndDeploymentsWrites: Writes[AppDefinition.WithTaskCountsAndDeployments] = Writes { app =>
-    val appJson = AppDefinitionWrites.writes(app).as[JsObject]
-
-    appJson ++ Json.obj(
-      "tasksStaged" -> app.tasksStaged,
-      "tasksRunning" -> app.tasksRunning,
-      "tasksHealthy" -> app.tasksHealthy,
-      "tasksUnhealthy" -> app.tasksUnhealthy,
```
Comment (kolloch) [inline src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala]: Generally, I prefer this explicit conversion to the stuff using JsonProperties. I am bit confused about the pre-existing @JsonProperty methods, though. Was that redundant code here?

**Comment by kolloch** [comment]
Thanks for the patch! I am looking forward to separated API models!

When you change stuff: Please make changes in a different commit! I don't want to go through everything again ;)

* V2AppDefinition:
  * Long parameter lists: It's easy to confuse parameters. I'd prefer named parameters (there is an Intellij quick fix for that).
  * You converted some play json code into code using @JsonProperty annotations. I prefer the play json way.
* unused imports in V2AppDefinition/AppDefinition
* the validation package should be moved into the v2 package, since it is v2 specific
* Rename MarathonFacade into MarathonV2Facade?
* ModelValidation.checkGroup, I didn't get the TODO at first, other wording?

I don't like but its hard to change and this I agree with your decision to defer it (but please follow up!):

* MarathonSchedulerActor shouldn't reference V2AppDefinition
* The GroupManager uses v2.{BeanValidation, ModelValidation}

**Comment by ConnorDoyle** [comment]
@kolloch thanks for the review!  I'll make the suggested changes.

File: src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala (hunk 3 above)
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala]: _Why not a def?_ The value is lifted to a function object when passing it to the higher-order function anyway.
_Why?_ I found the first version confusing to read at first.

I'll change it back.

File: src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala (hunk 12 above)
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala]: These `Writes` still exist but have moved to `mesosphere.marathon.api.v2.json.Formats` with the other formatting helpers.

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala (hunk 4 above)
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: This was added by mistake, good catch :baseball:

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala (hunk 5 above)
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: Will make these `lazy` like the others.

File: src/main/scala/mesosphere/marathon/api/v2/json/V2AppDefinition.scala (hunk 6 above)
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/marathon/api/v2/json/V2AppDefinition.scala]: Not sure how that got there.  You're right of course, fixing.

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala (hunk 11 above)
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: This (and surroundings) moved to `V2AppDefinition`.

File: src/main/scala/mesosphere/marathon/api/v2/json/V2DeploymentPlan.scala (hunk 7 above)
Comment (ConnorDoyle) [inline src/main/scala/mesosphere/marathon/api/v2/json/V2DeploymentPlan.scala]: Fixed

**Comment by ConnorDoyle** [comment]
@kolloch: rebased and addressed review comments.

**Comment by kolloch** [comment]
:+1: Thanks a lot for the fast follow up to my nagging!

I am really looking forward to the next steps! Let's get them into 0.10.0.
